cmd/query/app/querysvc/v2/querysvc/service.go

```diff
@@ -200,10 +204,168 @@ func (qs QueryService) receiveTraces(
 		seq(processTraces)
 	} else {
-		jptrace.AggregateTraces(seq)(func(trace ptrace.Traces, err error) bool {
-			return processTraces([]ptrace.Traces{trace}, err)
-		})
+		if qs.options.MaxTraceSize > 0 {
+			qs.aggregateTracesWithLimit(seq)(func(trace ptrace.Traces, err error) bool { // apply max trace size limit
+				return processTraces([]ptrace.Traces{trace}, err)
+			})
+		} else {
+			jptrace.AggregateTraces(seq)(func(trace ptrace.Traces, err error) bool {
+				return processTraces([]ptrace.Traces{trace}, err)
+			})
+		}
 	}
 
 	return foundTraceIDs, proceed
 }
+
+// It stops aggregating once the limit is reached for a trace and adds a warning.
+func (qs QueryService) aggregateTracesWithLimit(
+	tracesSeq iter.Seq2[[]ptrace.Traces, error],
+) iter.Seq2[ptrace.Traces, error] {
+	return func(yield func(trace ptrace.Traces, err error) bool) {
+		currentTrace := ptrace.NewTraces()
+		currentTraceID := pcommon.NewTraceIDEmpty()
+		spanCount := 0
+		truncated := false
+		skipCurrentTrace := false
+
+		tracesSeq(func(traces []ptrace.Traces, err error) bool {
+			if err != nil {
+				yield(ptrace.NewTraces(), err)
+				return false
+			}
+			for _, trace := range traces {
+				resources := trace.ResourceSpans()
+				if resources.Len() == 0 {
+					continue
+				}
+
+				scopes := resources.At(0).ScopeSpans()
+				if scopes.Len() == 0 {
+					continue
+				}
+
+				spans := scopes.At(0).Spans()
+				if spans.Len() == 0 {
+					continue
+				}
+
+				traceID := spans.At(0).TraceID()
+
+				if currentTraceID != traceID {
+					if currentTrace.ResourceSpans().Len() > 0 {
+						if truncated {
+							qs.addTruncationWarning(currentTrace)
+						}
+						if !yield(currentTrace, nil) {
+							return false
+						}
+					}
+					currentTrace = ptrace.NewTraces()
+					currentTraceID = traceID
+					spanCount = 0
+					truncated = false
+					skipCurrentTrace = false
+				}
+
+				// If already truncated, skip remaining
+				if skipCurrentTrace {
+					continue
+				}
+
+				// Count spans in incoming trace
+				incomingSpanCount := 0
+				for i := 0; i < resources.Len(); i++ {
+					res := resources.At(i)
+					for j := 0; j < res.ScopeSpans().Len(); j++ {
+						incomingSpanCount += res.ScopeSpans().At(j).Spans().Len()
+					}
+				}
+
+				// Check if adding these spans would exceed the limit
+				if spanCount+incomingSpanCount > qs.options.MaxTraceSize {
+					remaining := qs.options.MaxTraceSize - spanCount
+					if remaining > 0 {
+						qs.copySpansUpToLimit(trace, currentTrace, remaining)
+						spanCount = qs.options.MaxTraceSize
+					}
+					truncated = true
+					skipCurrentTrace = true
+					continue
+				}
+
+				// add all spans from this batch
+				mergeTraces(trace, currentTrace)
+				spanCount += incomingSpanCount
+			}
+			return true
+		})
+
+		// for last trace if it exists
+		if currentTrace.ResourceSpans().Len() > 0 {
+			if truncated {
+				qs.addTruncationWarning(currentTrace)
+			}
+			yield(currentTrace, nil)
+		}
+	}
+}
```

Comment:
this function is 100 LOC. The corresponding jtrace.AggregateTraces() is 30 LOC. I recommend:
  * rename `jtrace.AggregateTraces()` to `jtrace.AggregateTracesWithLimit()` that takes max size arg
  * add `jtrace.AggregateTraces()` that invokes `AggregateTracesWithLimit` with limit=0
  * deal with limit in the mergeTraces function